File: 프로그래머스/lv2/68645. 삼각 달팽이/삼각 달팽이.py
def solution(n):
    # 하, 우 좌상향 방향 순서
    dr = [1, 0, -1]
    dc = [0, 1, -1]
    answer = []
    
    # 삼각형을 왼쪽으로 붙여 직각삼각형처럼 본다
    triangle = [[0] * (i + 1) for i in range(n)]

    # 1 시작점 및 방향 d 초기화
    r, c = 0, 0
    triangle[r][c] = 1
    d = 0
    flag = False
    while True:
        nr, nc = r + dr[d], c + dc[d]
        # 다음칸이 인덱스가 맞는지, 비어있는지 확인
        if 0 <= nr < n and 0 <= nc <= nr and triangle[nr][nc] == 0:
            triangle[nr][nc] = triangle[r][c] + 1
            r, c = nr, nc
            flag = False

        # 유효하지 않은칸이면 방향바꿔서 다시 확인    
        else:
            # 이전 루프에서 한번 방향을 바꿧는데 또 유효하지 않으면 더이상 채울곳이 없다
            if flag:
                break
            d = (d + 1) % 3
            flag = True
            
    
    for line in triangle:
        answer += line
        
    return answer

File: 프로그래머스/2/68645. 삼각 달팽이/삼각 달팽이.java
import java.util.*;

class Solution {
    static int[] dr = {1, 0, -1};
    static int[] dc = {0, 1, -1};
    
    public int[] solution(int n) {
        int[] answer = new int[n * (1 + n) / 2];
        
        int[][] triangle = new int[n][n];
        
        Deque<int[]> dq = new ArrayDeque<int[]>();
        dq.offer(new int[]{0, 0});
        triangle[0][0] = 1;
        int num = 2;
        int d = 0;
        
        while (!dq.isEmpty()) {
            int[] pos = dq.poll();
            int nr = pos[0] + dr[d], nc = pos[1] + dc[d];
            if (0 <= nr && nr < n && 0 <= nc && nc < n && triangle[nr][nc] == 0) {
                triangle[nr][nc] = num++;
                dq.offer(new int[]{nr, nc});
            } else {
                d = (d + 1) % 3;
                nr = pos[0] + dr[d];
                nc = pos[1] + dc[d];
                if (0 <= nr && nr < n && 0 <= nc && nc < n && triangle[nr][nc] == 0) {
                    triangle[nr][nc] = num++;
                    dq.offer(new int[]{nr, nc});
                }
            }
        }
        
        int idx = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < i + 1; j++) {
                answer[idx] = triangle[i][j];
                idx++;
            }
        }
        
        return answer;
    }
}